### PR TITLE
Unchecked integer arithmetic in program-runtime-crate

### PR DESCRIPTION
**Pull request by Lichtso**
#### Problem
Crate wide `#![allow(clippy::integer_arithmetic)]` in the program-runtime.

#### Summary of Changes
Replaces unchecked integer arithmetic by guarded versions. 

Fixes #

### CODE REVIEW

File: program-runtime/src/invoke_context.rs [hunk 1]
```diff
@@ -315,13 +316,17 @@ impl<'a> InvokeContext for ThisInvokeContext<'a> {
...
                         "failed to verify account {}: {}",
-                        self.pre_accounts[unique_index].key(),
+                        pre_account.key(),
                         err
                     );
                     err
                 })?;
-            pre_sum += u128::from(self.pre_accounts[unique_index].lamports());
-            post_sum += u128::from(account.lamports());
+            pre_sum = pre_sum
```
Comment (Lichtso) [inline program-runtime/src/invoke_context.rs:324]: Critical changes here ...

File: program-runtime/src/invoke_context.rs [hunk 2]
```diff
@@ -383,8 +388,12 @@ impl<'a> InvokeContext for ThisInvokeContext<'a> {
                                 ic_logger_msg!(logger, "failed to verify account {}: {}", key, err);
                                 err
                             })?;
-                        pre_sum += u128::from(pre_account.lamports());
-                        post_sum += u128::from(account.lamports());
+                        pre_sum = pre_sum
```
Comment (Lichtso) [inline program-runtime/src/invoke_context.rs:391]: ... and here.

**Comment by jackcmay** [comment]
Was there an issue or test case that highlights the problem?  

If pre or post sum saturates we have problems, maybe checked with a failure instead if you are worried about this math?

**Comment by Lichtso** [comment]
We don't have test cases for the u128 overflow because they won't.
However it is still nice to get rid of the crate wide clippy rule.

`pre_sum` and `post_sum` deliberately don't use `saturating_add` but `checked_add`.

**Pull request merged**